### PR TITLE
feat: add check key type for nacl.Encrypter

### LINKED ISSUE #359: Check the private key type used when creating a nacl.NewEncrypter
**Is your feature request related to a problem? Please describe.**
With the addition of different public key types, it is possible to create a signer with an incompatible public key

**Describe the solution you'd like**
Check public key type is valid before creating signer, currently it must be `ed25519`

**Describe alternatives you've considered**
N/a

**Additional context**
`crypto/cipher/nacl/encrypter.go` NewEncrypter method should check is key is compatible then return error if not

### PR DESCRIPTION
**Pull request by Despire**
nacl.NewEncrypter now checks if key type is valid as described in #359 

Fixes #359

### CODE REVIEW

**Comment by Despire** [comment]
@robdefeo Is this ok, or do you suggest any changes ?

**Comment by robdefeo** [comment]
@Despire there is an `Encrypter` interface in `crypto/cipher/cipher.go`. Your implementation of `nacl` matches closer to the `Decrypter` interface. In many ways this makes more sense, it makes a good case that the interface should be changed. Can you take a look at the interfaces, implementations, and usage. I'd like to get your thoughts on this also. Then I can raise some issues. Thanks

**Comment by Despire** [comment]
@robdefeo hm.. haven't noticed the `Encrypter` interface. Thanks for pointing it out to me.
With the current `Encrypter` interface, the encrypters (`nacl`, `aes256cbc`...) are kind of "General" (i.e. the same Encrypter instance can be used for multiple public keys), however i didn't found any use of this in the codebase. 
For example the handler [SendMessage](https://github.com/mailchain/mailchain/blob/master/cmd/mailchain/internal/http/handlers/send_message.go#L43) function which uses the `aes256cbc` encrypter, uses it only with one public key which is passed in the request body. Also in [nacl.Encrypter](https://github.com/mailchain/mailchain/blob/master/crypto/cipher/nacl/encrypter.go#L20), we could pass multiple keys into [nacl.Encrypter.Encrypt](https://github.com/mailchain/mailchain/blob/master/crypto/cipher/nacl/encrypter.go#L24) and the same "randomness source" would be used with multiple keys.

From my point of view, it makes more sense to "attach" a single key with a `Encrypter` instance, since it is used in that sense anyways. (but i could be missing some information where the <b>multiple keys per one instance </b> could be used). Let me know Thanks

**Comment by Despire** [comment]
Hey @robdefeo 
I've prepared this PR for the refactor of the `cipher.Encrypter` interface. I think this change would make more sense (As there is no use-case in the codebase that uses the same encrypter with multiple keys as I have described in the previous comment). Let me know what you think, Thanks !

File: crypto/cipher/nacl/encrypter.go [hunk 1]
```diff
@@ -10,30 +10,31 @@ import (
...
+// Encrypt encrypts the message with the key that was attached to it.
+func (e Encrypter) Encrypt(message cipher.PlainContent) (cipher.EncryptedContent, error) {
+	encrypted, err := easySeal(message, e.publicKey.Bytes(), e.rand)
 	return bytesEncode(encrypted), err
 }
 
 func validatePublicKeyType(recipientPublicKey crypto.PublicKey) error {
 	switch recipientPublicKey.(type) {
-	case ed25519.PublicKey:
+	case ed25519.PublicKey, *ed25519.PublicKey:
```
Comment (robdefeo) [inline crypto/cipher/nacl/encrypter.go]: Good spot 👍

File: crypto/cipher/nacl/encrypter.go [hunk 2]
```diff
@@ -10,30 +10,31 @@ import (
...
-func NewEncrypter() Encrypter {
-	return Encrypter{rand: rand.Reader}
+// NewEncrypter creates a new encrypter with crypto rand for reader,
+// and attaching the public key to the encrypter.
+func NewEncrypter(publicKey crypto.PublicKey) (*Encrypter, error) {
+	if err := validatePublicKeyType(publicKey); err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	return &Encrypter{rand: rand.Reader, publicKey: publicKey}, nil
```
Comment (robdefeo) [inline crypto/cipher/nacl/encrypter.go]: Clearer 👍

File: internal/envelope/0x01.go [hunk 3]
```diff
@@ -20,14 +20,13 @@ import (
...
 
-	"github.com/mailchain/mailchain/crypto"
 	"github.com/mailchain/mailchain/crypto/cipher"
 	"github.com/pkg/errors"
 )
 
 // NewZeroX01 creates a new envelope of type ZeroX01.
 // ZeroX01 envelope allows sending private messages with the minimal bytes by using `Uint64Bytes`.
-func NewZeroX01(encrypter cipher.Encrypter, pubkey crypto.PublicKey, opts *CreateOpts) (*ZeroX01, error) {
+func NewZeroX01(encrypter cipher.Encrypter, opts *CreateOpts) (*ZeroX01, error) {
```
Comment (robdefeo) [inline internal/envelope/0x01.go]: Another place where your change looks good is on the envelopes

File: internal/mailbox/message.go [hunk 4]
```diff
@@ -67,7 +67,7 @@ func SendMessage(ctx context.Context, protocol, network string, msg *mail.Messag
 		locOpt,
 	}
 
-	env, err := envelope.NewEnvelope(encrypter, pubkey, opts)
+	env, err := envelope.NewEnvelope(encrypter, opts)
```
Comment (robdefeo) [inline internal/mailbox/message.go]: Looks much better like this 👍

File: crypto/cipher/cipher.go [hunk 5]
```diff
@@ -56,5 +52,5 @@ type Decrypter interface {
 // Returned encrypted data must include what encryption method was used as the first byte.
 // The data can be decrypted using the corresponding PrivateKey and Decrypter method.
 type Encrypter interface {
-	Encrypt(pubKey crypto.PublicKey, plain PlainContent) (EncryptedContent, error)
+	Encrypt(PlainContent) (EncryptedContent, error)
```
Comment (robdefeo) [inline crypto/cipher/cipher.go]: This is a simple but great change. I like how it makes the interface easier to understand and there is no current use case for using the same encrypter with different public keys.

**Pull request merged**